language-plutus-core/src/Language/PlutusCore/Constant/Function.hs

```diff
@@ -31,6 +32,11 @@ typeSchemeToType (TypeSchemeAllType proxy schK) = case proxy of
             a    = TyName $ Name text $ Unique uniq
         in TyForall () a (Type ()) $ typeSchemeToType (schK Proxy)
 
+countArgs :: TypeScheme uni as r -> Int
+countArgs (TypeSchemeResult _)       = 0
+countArgs (TypeSchemeArrow _ schB)   = 1 + countArgs schB
+countArgs (TypeSchemeAllType _ schK) = countArgs (schK Proxy)
```

Comment:
I think I mentioned this earlier somewhere: what if we count type arguments as well and then decrement the counter on encountering them in the machine? So that

```
f : integer -> all a. a
```

is handled correctly and does not evaluate before the final type argument is supplied.

It should also help us with Untyped Plutus Core where type arguments become something else that has to be taken into account as well.

Reply:
Yes, that makes sense.  Having thought a bit more, it probably doesn't matter too much if we do something inefficient in the typed CEK machine: if we do keep it it once we have untyped PLC it's not critical that it's super-efficient.